onnxruntime/core/providers/webgpu/string_utils.h

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <absl/strings/internal/ostringstream.h>
+
+namespace onnxruntime {
+namespace webgpu {
+
+constexpr const size_t kStringInitialSizeSetByOffsetImpl = 128;
+constexpr const size_t kStringInitialSizeGetByOffsetImpl = 128;
+constexpr const size_t kStringInitialSizeShaderSourceCode = 2048;
+#ifndef NDEBUG
+constexpr const size_t kStringInitialSizeCacheKey = 512;
+#else
+constexpr const size_t kStringInitialSizeCacheKey = 256;
+#endif
+
+using OStringStream = absl::strings_internal::OStringStream;
+
+namespace detail {
+inline void OStringStreamAppendImpl(std::ostream& /*ss*/) noexcept {
+}
+
+template <typename T>
+inline void OStringStreamAppendImpl(std::ostream& ss, const T& t) noexcept {
+  ss << t;
+}
+
+template <typename T, typename... Args>
+inline void OStringStreamAppendImpl(std::ostream& ss, const T& t, const Args&... args) noexcept {
+  OStringStreamAppendImpl(ss, t);
+  OStringStreamAppendImpl(ss, args...);
+}
```

Comment:
Can we add something like this to decay `const char[]` to const char*? It saves significant binary size to do so.

https://github.com/microsoft/onnxruntime/blob/1bda91fc57d61367d245db808106687b43036bd1/include/onnxruntime/core/common/make_string.h#L91-L102